File: scripts/xhs/direct_message.js
// 只通过网页控件发送，不调用私有发送 API。与 Python 共享一组 DOM 状态判据。
(params) => {
    const visible = el => !!el && el.getBoundingClientRect().width > 0 &&
        el.getBoundingClientRect().height > 0 && getComputedStyle(el).visibility !== 'hidden';
    const one = selector => {
        const nodes = [...document.querySelectorAll(selector)].filter(visible);
        return nodes.length === 1 ? nodes[0] : null;
    };
    // 使用纯文本，保留换行；已有网页表情按其 alt 文本读取，避免漏掉草稿内容。
    const plainText = el => {
        if (!el) return '';
        const copy = el.cloneNode(true);
        copy.querySelectorAll('img').forEach(img => img.replaceWith(img.alt || '\uFFFC'));
        copy.querySelectorAll('br').forEach(br => br.replaceWith('\n'));
        copy.querySelectorAll('div,p').forEach(block => block.prepend('\n'));
        return copy.textContent.replace(/\r\n/g, '\n').trim();
    };
    const onChat = location.origin === 'https://www.xiaohongshu.com' &&
        /^\/chat(?:\/|$)/.test(location.pathname);
    const editor = one('.xhs-im-chat-window .xhs-im-input-bar-editor');
    const active = one('.xhs-im-conv-item--active[data-conv-kind="c2c"]');
    const header = one('.xhs-im-chat-window__header-name');
    const pathId = location.pathname.match(/^\/chat\/([0-9a-fA-F]{24})\/?$/)?.[1] || '';
    const outgoing = () => [...document.querySelectorAll('.xhs-im-msg-list .chat-item')]
        .filter(el => el.querySelector('.chat-item__bubble--me'))
        .map(el => ({
            message_id: el.getAttribute('data-message-id') || '',
            store_id: el.getAttribute('data-store-id') || '',
            text: plainText(el.querySelector('.xhs-im-bubble__text')),
            failed: !!el.querySelector('.chat-item__status-btn--failed'),
            pending: !!el.querySelector('.chat-item__status-loading'),
        }));
    const state = () => ({
        on_chat_page: onChat,
        ready: onChat && !!one('.xhs-im-view'),
        conversation_ready: !!editor && !!header && !!active &&
            active.getAttribute('data-conv-id') === pathId,
        user_id: pathId,
        recipient: header?.textContent.trim() || '',
        draft: plainText(editor),
        outgoing: params.action === 'check' ? outgoing() : [],
        not_logged_in: visible(document.querySelector('.login-container')),
    });
    if (params.action === 'snapshot') return state();
    if (params.action === 'list') {
        if (!onChat) return {error: '当前不是小红书消息页面'};
        const conversations = [...document.querySelectorAll(
            '.xhs-im-conv-item[data-conv-kind="c2c"]'
        )].map(el => ({
            user_id: el.getAttribute('data-conv-id'),
            nickname: el.querySelector('.xhs-im-conv-item__name')?.textContent.trim() || '',
        })).filter(item => !params.name || item.nickname === params.name);
        return {conversations};
    }
    const check = () => {
        if (!onChat || !pathId || pathId !== params.user_id || !active ||
            active.getAttribute('data-conv-id') !== params.user_id ||
            header?.textContent.trim() !== params.expected_name ||
            active.querySelector('.xhs-im-conv-item__name')?.textContent.trim() !==
                params.expected_name) return '收件人 ID、单人会话和昵称未同时匹配，已停止';
        if (!editor || editor.getAttribute('contenteditable') !== 'true' ||
            editor.getAttribute('aria-disabled') === 'true' ||
            !editor.getAttribute('data-placeholder')) return '私信输入框不可用，可能被限制发送';
        if (document.querySelector('.xhs-im-input-bar-wrap .xhs-im-input-bar-quote')) {
            return '输入框附带引用，请先清除引用后再发送纯文字私信';
        }
        return '';
    };
    const error = check();
    if (error) return {error};
    if (params.action === 'check') return state();
    if (params.action === 'fill') {
        const previous = plainText(editor);
        if (previous && previous !== params.content) return {error: '已有不同的草稿，未覆盖或发送'};
        editor.focus();
        editor.textContent = params.content;
        editor.dispatchEvent(new InputEvent('input', {bubbles: true, inputType: 'insertText'}));
        return {filled: true};
    }
    if (params.action === 'send') {
        if (plainText(editor) !== params.content) return {error: '发送前正文发生变化，已停止'};
        const messages = outgoing();
        if (messages.at(-1)?.text === params.content) return {error: '检测到重复私信，已停止'};
        const before_ids = messages.map(message => message.message_id);
        editor.focus();
        // 网页桌面端的发送入口为 Enter；Shift+Enter 为换行。
        editor.dispatchEvent(new KeyboardEvent('keydown', {
            key: 'Enter', code: 'Enter', keyCode: 13, which: 13,
            bubbles: true, cancelable: true, shiftKey: false, isComposing: false,
        }));
        return {submitted: true, before_ids};
    }
    return {error: '未知私信操作'};
}

File: scripts/xhs/direct_message.py
"""网页文字私信：核对收件人、填写预览、单次发送并等待服务端消息标识。"""

from __future__ import annotations

import json
import re
import time
from pathlib import Path
from typing import Any

from .bridge import BridgePage
from .errors import NotLoggedInError, XHSError

CHAT_URL = "https://www.xiaohongshu.com/chat"
MAX_TEXT_LENGTH = 1000  # 网页按 JavaScript UTF-16 length 限制输入长度。
_SCRIPT = Path(__file__).with_name("direct_message.js").read_text(encoding="utf-8")


class DirectMessageError(XHSError):
    """私信发送前校验失败。"""


def read_message_file(path: str) -> str:
    """在连接浏览器前校验文件和文字，防止网页静默截断。"""
    file = Path(path)
    if not file.is_absolute():
        raise DirectMessageError("私信正文文件必须使用绝对路径")
    try:
        content = file.read_text(encoding="utf-8-sig").replace("\r\n", "\n").strip()
    except (OSError, UnicodeError) as exc:
        raise DirectMessageError("无法读取 UTF-8 私信正文文件") from exc
    validate_content(content)
    return content


def validate_content(content: str) -> None:
    if not content.strip():
        raise DirectMessageError("私信正文不可为空")
    if len(content.encode("utf-16-le")) // 2 > MAX_TEXT_LENGTH:
        raise DirectMessageError("私信正文超过网页 1000 字符限制（表情可能占两个字符）")


def _run(page: BridgePage, action: str, **params: Any) -> dict:
    result = page.evaluate(f"({_SCRIPT})({json.dumps({'action': action, **params})})")
    if not isinstance(result, dict):
        raise DirectMessageError("无法读取私信页面状态")
    if result.get("not_logged_in"):
        raise NotLoggedInError()
    if result.get("error") or result.get("__xhs_error"):
        raise DirectMessageError(result.get("error") or result["__xhs_error"])
    return result


def _wait_ready(page: BridgePage, user_id: str = "", timeout: float = 45.0) -> dict:
    deadline = time.monotonic() + timeout
    while True:
        state = _run(page, "snapshot")
        if state.get("ready") and (
            not user_id or (state.get("user_id") == user_id and state.get("conversation_ready"))
        ):
            return state
        if time.monotonic() >= deadline:
            raise DirectMessageError("私信页面未就绪，请检查登录、网页消息功能及收件人是否可访问")
        time.sleep(0.5)


def list_conversations(page: BridgePage, name: str = "") -> dict:
    """只列出网页当前已加载的单人会话，不读取消息正文。"""
    state = _run(page, "snapshot")
    if not state.get("on_chat_page"):
        page.navigate(CHAT_URL)
        page.wait_for_load()
    _wait_ready(page)
    result = _run(page, "list", name=name)
    return {"success": True, "scope": "loaded_conversations", **result}


def _open_recipient(page: BridgePage, user_id: str, expected_name: str) -> dict:
    if not re.fullmatch(r"[0-9a-fA-F]{24}", user_id):
        raise DirectMessageError("用户 ID 必须是主页或会话中的 24 位十六进制 ID")
    if not expected_name.strip():
        raise DirectMessageError("必须提供收件人的完整昵称用于交叉核对")
    state = _run(page, "snapshot")
    if state.get("user_id") != user_id:
        if state.get("draft"):
            raise DirectMessageError("当前会话存在草稿，请先处理，避免切换会话丢失内容")
        page.navigate(f"{CHAT_URL}/{user_id}")
        page.wait_for_load()
    _wait_ready(page, user_id)
    return _run(page, "check", user_id=user_id, expected_name=expected_name)


def fill_direct_message(page: BridgePage, user_id: str, expected_name: str, content: str) -> dict:
    """填写文字并返回预览，不触发发送。不同的已有草稿不会被覆盖。"""
    validate_content(content)
    _open_recipient(page, user_id, expected_name)
    _run(page, "fill", user_id=user_id, expected_name=expected_name, content=content)
    # Vue 的 input 处理可能异步更新；再次读取确认没有截断或切换会话。
    state = _run(page, "check", user_id=user_id, expected_name=expected_name)
    if state.get("draft") != content:
        raise DirectMessageError("输入框正文与预期不一致，未发送")
    return {
        "success": True,
        "status": "draft",
        "sent": False,
        "user_id": user_id,
        "recipient": expected_name,
        "content": content,
    }


def _is_acknowledged(message: dict) -> bool:
    store_id = str(message.get("store_id", ""))
    return (
        bool(message.get("message_id"))
        and store_id.isdigit()
        and int(store_id) > 0
        and not message.get("failed")
        and not message.get("pending")
    )


def send_direct_message(
    page: BridgePage,
    user_id: str,
    expected_name: str,
    content: str,
    *,
    confirmed: bool = False,
    timeout: float = 20.0,
) -> dict:
    """发送一次；发送后的异常、失败或超时均不自动重发。"""
    if not confirmed:
        raise DirectMessageError("发送私信需要 --confirm；仅预览请用 fill-direct-message")
    validate_content(content)
    state = _open_recipient(page, user_id, expected_name)
    outgoing = state.get("outgoing", [])
    if outgoing and outgoing[-1].get("text") == content:
        raise DirectMessageError("最近一条发出的私信与正文相同，已停止以避免重复发送；请核对会话")
    fill_direct_message(page, user_id, expected_name, content)
    base = {"user_id": user_id, "recipient": expected_name}
    # check 与 keydown 在同一次页面执行中完成，避免焦点切换后发给另一会话。
    try:
        submitted = _run(
            page, "send", user_id=user_id, expected_name=expected_name, content=content
        )
        before_ids = set(submitted["before_ids"])
        deadline = time.monotonic() + timeout
        while True:
            state = _run(page, "check", user_id=user_id, expected_name=expected_name)
            candidates = [
                message
                for message in state.get("outgoing", [])
                if message.get("message_id") not in before_ids and message.get("text") == content
            ]
            if len(candidates) == 1:
                message = candidates[0]
                if message.get("failed"):
                    return {
                        **base,
                        "success": False,
                        "status": "failed",
                        "error": "网页显示私信发送失败，未自动重试",
                        "message_id": message.get("message_id"),
                    }
                if _is_acknowledged(message) and not state.get("draft"):
                    return {
                        **base,
                        "success": True,
                        "status": "sent",
                        "message_id": message["message_id"],
                        "store_id": message["store_id"],
                        "verification": "server_message_id",
                    }
            if time.monotonic() >= deadline:
                break
            time.sleep(0.5)
    except Exception:
        # 桥接超时可能发生在网页已经发送之后，不把异常误报为安全可重试。
        return {
            **base,
            "success": False,
            "status": "unknown",
            "error": "发送期间连接或会话状态变化，结果未知；请核对会话，勿直接重试",
        }
    return {
        **base,
        "success": False,
        "status": "unknown",
        "error": "未在等待时间内确认服务端消息标识；请核对会话，勿直接重试",
    }
